cocos/base/ObjectFactory.cpp

```diff
@@ -47,7 +48,7 @@ ObjectFactory::TInfo::TInfo(const std::string& type, Instance ins)
 ObjectFactory::TInfo::TInfo(const std::string& type, InstanceFunc ins)
     :_class(type)
     ,_fun(nullptr)
-    ,_func(ins)
+    ,_func(std::move(ins))
```

Comment:
Why not use `const InstanceFunc &ins` instead of `std::move` here?

Reply:
Please read e.g.:
https://abseil.io/tips/117
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#exception-9

Personally I think that const& should be used here. This is what is recommended here: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f15-prefer-simple-and-conventional-ways-of-passing-information

And later if optimization is needed then pass by && or "perfect forwarding".

So personally I do not like "pass by value and then move" but some C++ experts and clang-tidy claim that's OK.

Change to "const&" will also fix this warning but there is another check for storing temporary variables which should be fixed with: 
"pass by &&" or "perfect forwarding" or "pass by value+std::move"

Reply:
> Change to "const&" will also fix this warning but there is another check for storing temporary variables which should be fixed with:
"pass by &&" or "perfect forwarding" or "pass by value+std::move"

I don't quite understand why this is a warning.

Reply:
InstanceFunc is std:function. std::function could be expensive to copy (e.g. when lamda capture big object  std::function do allocation).

Reply:
But it is copied when passing by value, isn't it?

Reply:
No, if you call like that:
```
f(std::move(instanceFun));
```

Reply:
It's copied once when passed to constructor. 
```cpp
ObjectFactory::TInfo::TInfo(const std::string& type, InstanceFunc ins)
```
It will be copied again in member intializer.

```cpp
,_func(ins)
```

I have a example here. https://wandbox.org/permlink/PXqmQmvCgdzvEFNF

Reply:
Here the same without copy but with moves:
https://wandbox.org/permlink/dW0KTViJhS9YNdnz

Reply:
@crazyhappygame yep, in your test case, it is not copied, but the invoker should pass using `std::move()`. Which means the invoker should be very careful to use it, if passing by reference, then invoker don't have to take care of it.

Reply:
Agree. I prefer pass by const & (and than pass by && if needed)

I think that there is another clang-tidy to check if std::move is used where needed https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html